File: LGPL/CommonSoftware/jacsalarm/src/alma/alarmsystem/source/ACSAlarmSystemInterfaceFactory.java
/*******************************************************************************
 * ALMA - Atacama Large Millimeter Array
 * Copyright (c) ESO - European Southern Observatory, 2011
 * (in the framework of the ALMA collaboration).
 * All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 *******************************************************************************/
package alma.alarmsystem.source;

import java.io.StringReader;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.util.logging.Logger;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import com.cosylab.CDB.DAL;

import alma.JavaContainerError.wrappers.AcsJContainerServicesEx;
import alma.acs.container.ContainerServicesBase;
import alma.acs.logging.AcsLogLevel;
import alma.acsErrTypeAlarmSourceFactory.ACSASFactoryNotInitedEx;
import alma.acsErrTypeAlarmSourceFactory.FaultStateCreationErrorEx;
import alma.acsErrTypeAlarmSourceFactory.SourceCreationErrorEx;
import alma.acsErrTypeAlarmSourceFactory.wrappers.AcsJACSASFactoryNotInitedEx;
import alma.acsErrTypeAlarmSourceFactory.wrappers.AcsJFaultStateCreationErrorEx;
import alma.acsErrTypeAlarmSourceFactory.wrappers.AcsJSourceCreationErrorEx;

/**
 * ACSAlarmSystemInterfaceFactory extends the CERN AlarmSystemInterfaceFactory
 * to create sources with different implementations depending on the actual
 * configuration in the CDB.
 * 
 * The type of implementation is in AlarmSystemConfiguration.xml: a property named
 * Implementation.
 * CERN implementation is used only if the property is CERN. 
 * If the property is not found, is ACS or the CDB record does not exist the 
 * ACS implementation for sources is used.
 * 
 * @author acaproni
 *
 */
public class ACSAlarmSystemInterfaceFactory {
	
	/**
	 * The path in the CDB of the AS configuraton
	 */
	private static final String CONFIGURATION_PATH="Alarms/Administrative/AlarmSystemConfiguration";
	
	/**
	 *  It is <code>true</code> if ACS implementation for sources must be used;
	 *  <code>false</code> means CERN implementation.
	 *  <P>
	 *  It is <code>null</code> if it has not yet been initialized.
	 */
	private static Boolean useACSAlarmSystem = null;
	
	/**
	 * At the present, while using the CERN implementation, 
	 * we use the same source for sending all the alarms
	 * so we can use a singleton and return the same object 
	 * to all the clients.
	 *  
	 *  @see AlarmSystemInterfaceProxy
	 */
	private static ACSAlarmSystemInterface source=null;
	
	/**
	 * The logger
	 */
	private static Logger logger=null;
	
	/**
	 * Container services
	 */
	private static ContainerServicesBase containerServices;
	
	/**
	 * Init the static variables of the class
	 * This method has to be called before executing any other
	 * method. 
	 * 
	 * @param logger The logger
	 * @param dal The DAL to init the AS with CERN or ACS implementation
	 * @throws AcsJContainerServicesEx 
	 */
	public static void init(ContainerServicesBase containerServices) throws AcsJContainerServicesEx {
		if (containerServices==null) {
			throw new AcsJContainerServicesEx(new Exception("Invalid null ContainerServicesBase"));
		}
		ACSAlarmSystemInterfaceFactory.containerServices=containerServices;
		ACSAlarmSystemInterfaceFactory.logger = containerServices.getLogger();
		DAL dal = containerServices.getCDB();
		if (logger==null || dal==null) {
			throw new IllegalArgumentException("Invalid DAL or Logger from ContainerServicesBase");
		}
		
		useACSAlarmSystem = retrieveImplementationType(dal);
		if (logger!=null) {
			if (useACSAlarmSystem) {
				logger.log(AcsLogLevel.DEBUG,"Alarm system type: ACS");
			} else {
				logger.log(AcsLogLevel.DEBUG,"Alarm system type: CERN");
				try {
					initCmwMom();
				} catch (Throwable t) {
					throw new AcsJContainerServicesEx(new Exception("Error initing cmw-mom",t));
				}
			}
		}
	}
	
	/**
	 * Initialize cmw-mom (that in turn inits acs-jms) by setting the container services.
	 * 
	 * The container services must be set in a cmw-mom static variable in order the
	 * messages are published into the NC.
	 * The container services will be passed from this class down to the acs-jms classes.
	 * 
	 * @see cern.cmw.mom.pubsub.impl.ACSJMSTopicConnectionImpl
	 */
	private static void initCmwMom() throws Exception {
		if (containerServices==null) {
			throw new IllegalStateException("Trying to init cmw-mom with null ContainerServicesBase");
		}
		try {
			Thread t = Thread.currentThread();
			ClassLoader loader = t.getContextClassLoader();
			Class cl =loader.loadClass("cern.cmw.mom.pubsub.impl.ACSJMSTopicConnectionImpl");
			Field var = cl.getField("containerServices");
			var.set(null, containerServices);
			logger.log(AcsLogLevel.DEBUG,"cmw-mom/acs-jms initialized");
		} catch (Throwable t) {
			throw new Exception("Error setting ContainerServices into cmw-mom",t);
		}
	}
	
	/**
	 * Cleanup the class.
	 * This method has to be called outside of the class and performs all the necessary
	 * clean up
	 *
	 */
	public static void done() {
		useACSAlarmSystem=null;
	}
	
	/**
	 * Read the Implementation property from the Alarm System Configuration
	 * 
	 * @param dal The DAL
	 * 
	 * @return false if the Implementation property is CERN 
	 *         true otherwise
	 */
	private static boolean retrieveImplementationType(DAL dal) {
		if (dal==null) {
			return true;
		}
		String dao;
		try {
			dao = dal.get_DAO(CONFIGURATION_PATH);
		} catch (Exception e) {
			return true;
		}
		String implementation = getProperty(dao,"Implementation");
		return implementation==null || !implementation.equals("CERN");
	}
	
	/**
	 * Get the value of a property from the DAO.
	 * 
	 * @param dao The dao (XML string)
	 * @param propName The name of the property
	 * @return The value of the property with the given name
	 *         null if the property doesn't exist
	 */
	private static String getProperty(String dao, String propName) {
		if (dao==null || propName==null) {
			return null;
		}
		DocumentBuilder builder = null;
		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
		
		try {
			builder = factory.newDocumentBuilder();
		} catch (Exception e) {
			System.out.println("Error instantiating the document builder");
			System.out.println(e.getMessage());
			e.printStackTrace();
			return null;
		}
		Document doc;
		try {
			doc = builder.parse(new InputSource(new StringReader(dao)));
		} catch (Exception e) {
			System.out.println("Error parsing the DAO: ["+dao+"]");
			System.out.println(e.getMessage());
			e.printStackTrace();
			return null;
		}
		NodeList propNodeList = doc.getElementsByTagName("configuration-property");
		String val = null; // The value of the property to return
		for (int t=0; t<propNodeList.getLength(); t++) {
			Node node = propNodeList.item(t);
			// Get the attributes
			NamedNodeMap attributes = node.getAttributes();
			if (attributes==null) {
				continue;
			}
			// Get the attribute "name"
			Node attrNodeName=attributes.getNamedItem("name");
			if (attrNodeName==null) {
				continue;
			}
			String nameNode = attrNodeName.getNodeValue();
			if (nameNode==null) {
				continue;
			}
			if (nameNode.equals(propName)) {
				// We have found the property!
				Node child = node.getFirstChild();
				val = child.getNodeValue();
				break;
			}
			
		}
		return val;
	}
	
	/**
	 * Create a new instance of an alarm system interface.
	 * @param sourceName the source name.
	 * @return the interface instance.
	 * @throws ASIException if the AlarmSystemInterface instance can not be created.
	 */
	  public synchronized static ACSAlarmSystemInterface createSource(String sourceName) throws ACSASFactoryNotInitedEx, SourceCreationErrorEx  {
		  if (useACSAlarmSystem==null) {
			  Exception e = new IllegalStateException("Factory not initialised");
			  throw new AcsJACSASFactoryNotInitedEx(e).toACSASFactoryNotInitedEx();
		  }
		  if (useACSAlarmSystem) {
			  return new ACSAlarmSystemInterfaceProxy(sourceName,logger);
		  } else {
			  if (source!=null) {
				  return source;
			  }
			  try {
				  Thread t = Thread.currentThread();
				  ClassLoader loader = t.getContextClassLoader();
				  Class cl =loader.loadClass("alma.acs.alarmsystem.binding.ACSLaserSource");
				  Class[] classes = {String.class , Logger.class };
				  Constructor constructor = cl.getConstructor(classes);

				  // TODO: take ContainerServicesBase object received in the init method, 
				  // and set it on ACSJMSTopicConnectionImpl.containerServices
				  // Of course in after the ACS 7.0 release rush, this static field communication 
				  // must be replaced with with real calls, or at least the absence of such an object must be detected early.
				  source=(ACSAlarmSystemInterface)constructor.newInstance(sourceName,logger);
				  return source;
			  } catch (Throwable t) {
				  System.out.println("ERROR: "+t.getMessage());
				  t.printStackTrace();
				  throw new AcsJSourceCreationErrorEx(t).toSourceCreationErrorEx();
			  }
		  }
	  }

	  /**
	   * Create a new instance of an alarm system interface without binding it to any source.
	   * @return the interface instance.
	  * @throws ASIException if the AlarmSystemInterface instance can not be created.
	   */
	  public static ACSAlarmSystemInterface createSource() throws ACSASFactoryNotInitedEx, SourceCreationErrorEx {
		  return createSource("UNDEFINED");
	  }
	  
	  /** 
	   * Factory method for creating ACSFaultState instances.
	   * @return a new ACSFaultState instance.
	   *
	   */
	  public synchronized static ACSFaultState createFaultState() throws ACSASFactoryNotInitedEx, FaultStateCreationErrorEx{
		  if (useACSAlarmSystem==null) {
			  Exception e = new IllegalStateException("Factory not initialised");
			  throw new AcsJACSASFactoryNotInitedEx(e).toACSASFactoryNotInitedEx();
		  }
		  if (useACSAlarmSystem) {
			  return new ACSFaultStateImpl();
		  } else {
			  try {
				  Thread t = Thread.currentThread();
				  ClassLoader loader = t.getContextClassLoader();
				  Class cl =loader.loadClass("alma.acs.alarmsystem.binding.ACSLaserFaultStateImpl");
				  Class[] classes = {};
				  Constructor constructor = cl.getConstructor(classes);
				  return (ACSFaultState)constructor.newInstance();
			  } catch (Exception e) {
				  throw new AcsJFaultStateCreationErrorEx(e).toFaultStateCreationErrorEx();
			  }
		  }
	  }

	  /** 
	   * Factory method for creating ACSFaultState instances.
	   * @return a new ACSFaultState instance.
	   * @param family the fault family.
	   * @param member the fault member.
	   * @param code the fault code.
	   */
	  public static ACSFaultState createFaultState(String family, String member, int code) throws ACSASFactoryNotInitedEx, FaultStateCreationErrorEx {
		  ACSFaultState state = createFaultState();
		  state.setFamily(family);
		  state.setMember(member);
		  state.setCode(code);
		  return state;
	  }
	  
	  /**
	   * Return the type of AS used
	   * 
	   * @return True if ACS AS is used, false otherwise
	   * @throws IllegalStateException If the factory has not been initialized yet
	   */
	  public static boolean usingACSAlarmSystem() throws ACSASFactoryNotInitedEx {
		  if (useACSAlarmSystem==null) {
			  Exception e = new IllegalStateException("Factory not initialised");
			  throw new AcsJACSASFactoryNotInitedEx(e).toACSASFactoryNotInitedEx();
		  }
		  return ACSAlarmSystemInterfaceFactory.useACSAlarmSystem;
	  }
}

File: LGPL/CommonSoftware/acsalarm/src/ACSAlarmSystemInterfaceFactory.cpp
/*******************************************************************************
* ALMA - Atacama Large Millimiter Array
* (c) European Southern Observatory, 2006 
* 
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
* 
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
* 
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
*
* "@(#) $Id$"
*
* who       when      what
* --------  --------  ----------------------------------------------
* acaproni  2006-07-12  created 
*/

#include "vltPort.h"

#include "ConfigPropertyGetter.h"
#include "faultStateConstants.h"

static char *rcsId="@(#) $Id$"; 
static void *use_rcsId = ((void)&use_rcsId,(void *) &rcsId);

#include "ACSAlarmSystemInterfaceFactory.h"
#include "FaultState.h"
#include "asiConfigurationConstants.h"
#include <logging.h>

using namespace acsalarm;
using asiConfigurationConstants::ALARM_SOURCE_NAME;

bool* ACSAlarmSystemInterfaceFactory::m_useACSAlarmSystem = NULL;
maci::Manager_ptr ACSAlarmSystemInterfaceFactory::m_manager = maci::Manager::_nil();
AlarmSystemInterfaceFactory * ACSAlarmSystemInterfaceFactory::m_AlarmSystemInterfaceFactory_p = NULL;
void* ACSAlarmSystemInterfaceFactory::dllHandle = NULL;
ACE_Recursive_Thread_Mutex ACSAlarmSystemInterfaceFactory::main_mutex;
auto_ptr<acsalarm::AlarmSystemInterface> ACSAlarmSystemInterfaceFactory::sharedSource;

/**
 * Create a new instance of an alarm system interface without binding it to any source.
 */
auto_ptr<acsalarm::AlarmSystemInterface> ACSAlarmSystemInterfaceFactory::createSource() throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createSource entering");

	auto_ptr<acsalarm::AlarmSystemInterface> retVal;
	if (m_useACSAlarmSystem == NULL) {
		throw acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::createSource");
	}
	if (!(*m_useACSAlarmSystem)) {
		retVal = m_AlarmSystemInterfaceFactory_p->createSource();
	} else {
		retVal = createSource("UNDEFINED");
	}

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createSource exiting");
	return retVal;
}

/**
 * Create a new instance of an alarm system interface.
 */
auto_ptr<acsalarm::AlarmSystemInterface> ACSAlarmSystemInterfaceFactory::createSource(string sourceName) throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createSource(string) entering...");

	auto_ptr<acsalarm::AlarmSystemInterface> retVal;
	if (m_useACSAlarmSystem == NULL) {
		throw acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::createSource");
	}
	if (!(*m_useACSAlarmSystem)) {
		retVal = m_AlarmSystemInterfaceFactory_p->createSource(sourceName);
	} else {
		retVal.reset(new ACSAlarmSystemInterfaceProxy(sourceName));
	}

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createSource(string) exiting");
	return retVal;
}

/**
 * Getter for whether we're using the ACS Alarm system (true) or not (false).
 */
bool ACSAlarmSystemInterfaceFactory::usingACSAlarmSystem() throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{ 
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::usingACSAlarmSystem() entering");
	bool retVal = true;
	if(NULL == m_useACSAlarmSystem)
	{ 
		throw acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::usingACSAlarmSystem"); 
	}
	else	
	{
		retVal = *m_useACSAlarmSystem; 
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::usingACSAlarmSystem() exiting");
	return retVal;
}

auto_ptr<acsalarm::FaultState>ACSAlarmSystemInterfaceFactory::createFaultState(string family, string member, int code) 
	throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createFaultState(string, string, int) entering");
	auto_ptr<acsalarm::FaultState> retVal;
	if (m_useACSAlarmSystem==NULL) {
		throw acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::createFaultState(string, string, int)");
	}
	if (!(*m_useACSAlarmSystem)) {
		retVal = m_AlarmSystemInterfaceFactory_p->createFaultState(family, member, code);
	} else {
		retVal.reset(new acsalarm::FaultState(family, member, code));
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createFaultState(string, string, int) exiting");
	return retVal;
}
	
auto_ptr<acsalarm::FaultState>ACSAlarmSystemInterfaceFactory::createFaultState() throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createFaultState() entering");
	auto_ptr<acsalarm::FaultState> retVal;
	if (m_useACSAlarmSystem==NULL) {
		throw acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::createFaultState()");
	}
	if (!(*m_useACSAlarmSystem)) {
		retVal = m_AlarmSystemInterfaceFactory_p->createFaultState();
	} else {
		retVal.reset(new acsalarm::FaultState());
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createFaultState() exiting");
	return retVal;
}

maci::Manager_ptr ACSAlarmSystemInterfaceFactory::getManager()
{ 
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::getManager() entering/exiting");
	return m_manager; 
}

/**
 * Short-hand convenience API to create and send an alarm in a single step.
 */
void ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string & faultFamily, string & faultMember, int faultCode, bool active, string sourceName) 
	throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string, string, int, bool, string) entering");

	// create a Properties object and configure it, then assign to the FaultState
	Properties properties;

	ACSAlarmSystemInterfaceFactory::createAndSendAlarm(faultFamily, faultMember, faultCode, active, properties, sourceName);
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string, string, int, bool, string) exiting");
}

/**
 * Short-hand convenience API to create and send an alarm in a single step.
 */
void ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string & faultFamily, string & faultMember, int faultCode, bool active, Properties & faultProperties, string sourceName) 
	throw (acsErrTypeAlarmSourceFactory::ACSASFactoryNotInitedExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string, string, int, bool, Properties, string) entering");

	// create the FaultState
	auto_ptr<acsalarm::FaultState> fltstate = ACSAlarmSystemInterfaceFactory::createFaultState(faultFamily, faultMember, faultCode);

	// set the fault state's descriptor
	string stateString;
	if (active) 
	{
		stateString = faultState::ACTIVE_STRING;
	} 
	else 
	{
		stateString = faultState::TERMINATE_STRING;
	}
	fltstate->setDescriptor(stateString);
		
	// create a Timestamp and use it to configure the FaultState
	auto_ptr<Timestamp> tstampAutoPtr(new Timestamp());
	fltstate->setUserTimestamp(tstampAutoPtr);

	// create a Properties object (using copy constructor) and assign to the FaultState 
	Properties * propsPtr = new Properties(faultProperties);
	auto_ptr<Properties> propsAutoPtr(propsPtr);
	fltstate->setUserProperties(propsAutoPtr);

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if(NULL == sharedSource.get())
	{
		// Send the fault. We must use the "ALARM_SYSTEM_SOURCES" to match the name defined on CDB
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string, string, int, bool, Properties, string) creating shared source");
		sharedSource = ACSAlarmSystemInterfaceFactory::createSource(ALARM_SOURCE_NAME);
	}

	// push the FaultState using the source
	ACSAlarmSystemInterfaceFactory::sharedSource->push(*fltstate);

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::createAndSendAlarm(string, string, int, bool, Properties, string) exiting");
}

// called at shutdown by maciContainer
void ACSAlarmSystemInterfaceFactory::done() 
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::done entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	cleanUpAlarmSystemInterfacePtr();
	cleanUpSharedSource();
	cleanUpDLL();
	cleanUpBooleanPtr();
	cleanUpManagerReference();
	 
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::done exiting");
}

// private method called at shutdown
void ACSAlarmSystemInterfaceFactory::cleanUpManagerReference()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpManagerReference entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if (!CORBA::is_nil(m_manager)) {
		CORBA::release(m_manager);
		m_manager = maci::Manager::_nil();
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpManagerReference nilling manager reference");
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpManagerReference exiting");
}

// private method called at shutdown
void ACSAlarmSystemInterfaceFactory::cleanUpBooleanPtr()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpBooleanPtr entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if(NULL != m_useACSAlarmSystem) 
	{
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpBooleanPtr deleting m_useACSAlarmSystem ptr");
		delete m_useACSAlarmSystem;
		m_useACSAlarmSystem = NULL;
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpBooleanPtr exiting");
}

// private method called at shutdown
void ACSAlarmSystemInterfaceFactory::cleanUpDLL()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpDLL entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if(NULL != ACSAlarmSystemInterfaceFactory::dllHandle)
	{
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpDLL closing DLL");
		dlclose(ACSAlarmSystemInterfaceFactory::dllHandle);
		ACSAlarmSystemInterfaceFactory::dllHandle = NULL;
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpDLL exiting");
}

// private method called at shutdown
void ACSAlarmSystemInterfaceFactory::cleanUpSharedSource()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpSharedSource entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if(NULL != sharedSource.get())
	{
		// force the deletion of the allocated memory for the shared source auto_ptr
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpSharedSource nulling out shared source auto_ptr");
		sharedSource.reset();
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpSharedSource exiting");
}

// private method called at shutdown
void ACSAlarmSystemInterfaceFactory::cleanUpAlarmSystemInterfacePtr()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpAlarmSystemInterfacePtr entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if (NULL != m_useACSAlarmSystem && !(*m_useACSAlarmSystem) && NULL != m_AlarmSystemInterfaceFactory_p) {
		m_AlarmSystemInterfaceFactory_p->done();
		myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpAlarmSystemInterfacePtr deleting alarm system interface ptr");
		delete m_AlarmSystemInterfaceFactory_p;
		m_AlarmSystemInterfaceFactory_p = NULL;
	}
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::cleanUpAlarmSystemInterfacePtr exiting");
}

// public method: called at startup by maciContainer
bool ACSAlarmSystemInterfaceFactory::init(maci::Manager_ptr manager) throw (acsErrTypeAlarmSourceFactory::ErrorLoadingCERNDLLExImpl)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::init() entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	bool retVal = true;

	initImplementationType(manager);

	if (!(*m_useACSAlarmSystem)) 
	{
		retVal = initDLL();
	}

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::init() exiting");
	return retVal;
}

// private method callled during initialization (at container startup)
void ACSAlarmSystemInterfaceFactory::initImplementationType(maci::Manager_ptr manager)
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::initImplementationType entering");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	if (manager!=maci::Manager::_nil()) 
	{
		m_manager = maci::Manager::_duplicate(manager);
		m_useACSAlarmSystem = new bool(); // It implicitly says that the init has been called
		try
		{
			ConfigPropertyGetter pGetter(m_manager);
			string str = pGetter.getProperty("Implementation");
			*m_useACSAlarmSystem = !(str=="CERN");
		}
		catch(...)
		{
			// if we get any exception from accessing CDB we use the default ACS alarm system
			*m_useACSAlarmSystem=true;
		}
	} 
	else 
	{
		// if we were passed a NULL for the manager reference, this means we should use the ACS (logging) style for alarms
		// this typically will only happen in test code within the acsalarm module, which due to build/dependency order issues
		// cannot access things in the ACSLaser package (which is built later). 
		m_useACSAlarmSystem = new bool();
		*m_useACSAlarmSystem=true;
	}

	// Print a debug message
	if (*m_useACSAlarmSystem) {
		myLoggerSmartPtr->log(Logging::Logger::LM_DEBUG, "Using ACS alarm system");
	} else {
		myLoggerSmartPtr->log(Logging::Logger::LM_DEBUG, "Using CERN alarm system");
	}

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::initImplementationType exiting");
}

// private method callled during initialization (at container startup), if using CERN alarm style
bool ACSAlarmSystemInterfaceFactory::initDLL()
{
	Logging::Logger::LoggerSmartPtr myLoggerSmartPtr = getLogger();
	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::initDLL entering");
	bool retVal = true;

	myLoggerSmartPtr->log(Logging::Logger::LM_DEBUG, "ACSAlarmSystemInterfaceFactory::initDLL() loading CERN DLL...");

	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(main_mutex);
	// load the DLL and then set pointer m_AlarmSystemInterfaceFactory_p to point to the object
	// that is returned from the DLL's entry point function. From then on, we can use the pointer/object directly.
	ACSAlarmSystemInterfaceFactory::dllHandle = dlopen(CERN_ALARM_SYSTEM_DLL_PATH, RTLD_NOW|RTLD_GLOBAL);
	if(ACSAlarmSystemInterfaceFactory::dllHandle == NULL)
	{
		string errString = "ACSAlarmSystemInterfaceFactory::initDLL(): could not open DLL; error was:\n\n" + string(dlerror()); 
		myLoggerSmartPtr->log(Logging::Logger::LM_ERROR, errString);
		throw acsErrTypeAlarmSourceFactory::ErrorLoadingCERNDLLExImpl(__FILE__,__LINE__,"ACSAlarmSystemInterfaceFactory::initDLL");
	}
	// Call the well-defined entry point function of the DLL, to get an object 
	// which implements the AlarmSystemInterfaceFactory interface, which will be used for publishing 
	// CERN style alarms (i.e. alarms that go over the notification channel as opposed to just being logged)
	void * publisherFactoryFunctionPtr = dlsym(ACSAlarmSystemInterfaceFactory::dllHandle, CERN_ALARM_SYSTEM_DLL_FUNCTION_NAME);
	m_AlarmSystemInterfaceFactory_p = ((AlarmSystemInterfaceFactory*(*)())(publisherFactoryFunctionPtr))();
	myLoggerSmartPtr->log(Logging::Logger::LM_DEBUG, "ACSAlarmSystemInterfaceFactory::initDLL() successfully loaded DLL");
	retVal = m_AlarmSystemInterfaceFactory_p->init();

	myLoggerSmartPtr->log(Logging::Logger::LM_TRACE, "ACSAlarmSystemInterfaceFactory::initDLL() exiting.");
	return retVal;
}
